comms/dht/src/dedup/dedup_cache.rs

```diff
@@ -0,0 +1,184 @@
+use log::*;
+use std::cmp::max;
+use tari_crypto::tari_utilities::ByteArray;
+use tari_utilities::hex;
+
+const LOG_TARGET: &str = "comms::dht::dedup_cache";
+
+//#[derive(Clone, Debug, Queryable, Insertable, PartialEq)]
+#[derive(Clone, Debug, Queryable, Identifiable)]
+#[table_name = "dedup_cache"]
+struct DedupCacheSql {
+    pub id: i32,
+    pub body_hash: String,
+    pub number_of_hits: i32,
+    pub stored_at: NaiveDateTime,
+    pub last_hit_at: NaiveDateTime,
+}
+
+#[derive(Clone, Debug, Insertable, Default, AsChangeset)]
+#[table_name = "dedup_cache"]
+pub struct UpdateDedupCacheSql {
+    pub body_hash: Option<String>,
+    pub number_of_hits: Option<i32>,
+    pub last_hit_at: Option<NaiveDateTime>,
+}
+
+#[derive(Clone)]
+pub struct DedupCacheDatabase {
+    connection: DbConnection,
+    capacity: usize,
+}
+
+impl DedupCacheDatabase {
+    pub fn new(connection: DbConnection, capacity: usize) -> Self {
+        let capacity = max(capacity, 100);
+        debug!(
+            target: LOG_TARGET,
+            "Message dedup cache capacity initialized at {}", capacity,
+        );
+        Self { connection, capacity }
+    }
+
+    /// Inserts and returns Ok(true) if the item already existed and Ok(false) if it didn't
+    pub async fn insert_body_hash_if_unique(&self, body_hash: Vec<u8>) -> Result<bool, StorageError> {
+        let body_hash_string = hex::to_hex(&body_hash.as_bytes());
+        match self.insert_body_hash(body_hash_string.clone()).await {
+            Ok(val) => {
+                if val == 0 {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Unable to insert new entry into message dedup cache"
+                    );
+                }
+                Ok(false)
+            },
+            Err(e) => match e {
+                StorageError::UniqueViolation(_) => match self.update_number_of_hits(body_hash_string).await {
+                    Ok(_) => Ok(true),
+                    Err(e) => Err(e),
+                },
+                _ => Err(e),
+            },
+        }
+    }
+
+    /// Trims the dedup cache to the configured limit by removing the oldest entries
+    pub async fn truncate(&self) -> Result<usize, StorageError> {
+        let capacity = self.capacity;
+        self.connection
+            .with_connection_async(move |conn| {
+                let mut num_removed = 0;
+                let msg_count = dedup_cache::table
+                    .select(dsl::count(dedup_cache::id))
+                    .first::<i64>(conn)? as usize;
+                // Hysteresis added to minimize database impact
+                if msg_count > capacity {
+                    let remove_count = msg_count - capacity;
+                    let message_ids: Vec<i32> = dedup_cache::table
+                        .select(dedup_cache::id)
+                        .order_by(dedup_cache::stored_at.asc())
```

Comment:
To match the previous TtlCache behaviour, you should actually sort by `last_hit_at` instead

Reply:
Good catch

Reply:
The TTL principle will remove the oldest entries first, similar to sorting on `stored_at`. Using `last_hit`, a 24-hour old entry with a recent hit will remain in the cache.

Reply:
In the old TTL cache, when you insert a row, you give it a duration, when that expires, it is removed. When you re-insert an existing value, you insert the same duration, basically extending it's life, so if the duration was 24 hours and it was at 23 hours and a dup message came it, it would live another 24 hours.

```
                let already_exists = self
                    .msg_hash_cache
                    .insert(hash, (), self.config.msg_hash_cache_ttl)
 ```

Reply:
I concede about `order_by(dedup_cache::stored_at.asc())` being different than the original TTL.

My proposal here was actually that the oldest messages are always removed from the cache, not that a very old message with a recent hit gets an extended lifetime in the cache. This way part of a conversation can be replayed after the message's original stay in the cache expired if it was legitimate.

Changing to sort by `last_hit_at` as proposed.